### PR TITLE
Feat: 1640 add remaining gtfs features to the validator

### LINKED ISSUE #1640: Add remaining GTFS Features to the Validator
### Describe the problem

The spec team at MobilityData has started to structure the spec into features (e.g Wheelchair Accessibility, Pathways) to 1) make it easier for newcomers to the spec to understand what use cases can be addressed with it and 2) make it easy for governments and operators to require specific features in contracts and regulatory requirements. 

[In progress spreadsheet here](https://docs.google.com/spreadsheets/d/1kpbKOzlHtsJIPo3B4ABYu-Nxwvqutjcdcvgpa0wZvtA/edit#gid=1997312156)

We started adding these to the validator metadata report so the community can see what features are in different feeds and to also help with our own internal analytics. The first iteration of this list tracks **the presence of features** in the feed, as in if there is 1 occurrence, we count it. 

Now that the list of features and its logic are almost finalized, we need to 1) add new features that are missing in the validator and 2) do some minor change of logic. 

### Proposed solution

[FeedMetadata](https://github.com/MobilityData/gtfs-validator/blob/master/main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java)

Logic should check values so there has to be 1 record (not just the column headers) for each feature

- [x] ~~Change Bikes Allowed = if at least 1 value of bikes_allowed in trips.txt defined. Remove check for values 2 or 3.~~ (already done) 
- [x] Change Zone-Based Fares = if 1 record in areas.txt
      - [MBTA: Example feed](https://cdn.mbta.com/MBTA_GTFS.zip)
- [x] Change name of Frequency-Based Trips feature to Frequencies
    - [Bullrunner: Example feed download link](https://github.com/CUTR-at-USF/bullrunner-gtfs-realtime-generator/blob/master/bullrunner-gtfs.zip)
- [x] Add Route-Based Fares = If at least 1 value of network_id in routes.txt defined OR 1 record in networks.txt
      - [MBTA: Example feed](https://cdn.mbta.com/MBTA_GTFS.zip)
- [x] Add Time-Based Fares = If 1 record in timeframes.txt
- [x] Add Transfer Fares = If 1 record in fare_transfer_rules.txt
      - [MBTA: Example feed](https://cdn.mbta.com/MBTA_GTFS.zip)
- [x] Add Pathways (extra) = If 1 occurrence of max_slope OR min_width OR length OR stair_count in pathways.txt
      - [MBTA: Example feed](https://cdn.mbta.com/MBTA_GTFS.zip)
- [x] Add Levels = If 1 record in levels.txt
      - [MBTA: Example feed](https://cdn.mbta.com/MBTA_GTFS.zip)

### Alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by qcdyx**
**Summary:**

Added new GTFS features as described in #1640 

**Expected behavior:** 
MBTA.zip
image
Bullrunner.zip
image

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Add or update any needed [documentation](https://github.com/MobilityData/gtfs-validator/tree/master/docs) to the repo 
- [x] Format the title like "feat: [new feature short description]". Title must follow the Conventional Commit Specification(https://www.conventionalcommits.org/en/v1.0.0/).
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

**Review by emmambd — COMMENTED** [review]
@qcdyx Nice work! Two comments: 

- I don't see "Pathways (extra)" when I test with MBTA's feed
- I now get a `agency.txt	INVALID_HEADERS` error when I run the validator through the cli. This is making the agency column in the Summary report empty even when the agency is provided by the feed (which I can also see in your screenshot). I don't see this in our staging environment, so I'm wondering if this is introducing a new bug.

**Comment by emmambd** [comment]
@qcdyx A requirement change: Transfer Rules should be renamed "Transfer Fares"

**Comment by qcdyx** [comment]
image

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:230]: I think `hasAtLeastOneRecordForFields` will do an AND of all the specified functions that are passed to it ( hasMaxSlope, hasMinWidth, hasLength and hasStairCount)
But according to this it should be an OR:
```
Add Pathways (extra) = If 1 occurrence of max_slope OR min_width OR length OR stair_count in pathways.txt
```

**Comment by davidgamez** [comment]
> ❌ Invalid acceptance test. New Errors: 1422 out of 1476 datasets (~96%) are invalid due to code change, which is above the provided threshold of 1%. Dropped Errors: 19 out of 1476 datasets (~1%) are invalid due to code change, which is above the provided threshold of 1%. New Warnings: 6 out of 1476 datasets (~0%) are invalid due to code change, which is less than the provided threshold of 1%. Dropped Warnings: 67 out of 1476 datasets (~5%) are invalid due to code change, which is above the provided threshold of 1%. 3 out of 1479 sources (~0 %) are corrupted. Corrupted sources: br-minas-gerais-empresa-de-transportes-e-transito-de-belo-horizonte-bhtrans-gtfs-9 us-hawaii-thebus-gtfs-10 us-vermont-lake-champlain-ferries-gtfs-547 Commit: [314aebb](https://github.com/MobilityData/gtfs-validator/commit/314aebb4aa894b1e2bdd643ae5eab5656f885fa6) Download the full acceptance test report [here](https://github.com/MobilityData/gtfs-validator/actions/runs/7615360550) (report will disappear after 90 days). ❌ Invalid acceptance test.

I expected no changes in the acceptance tests as #1640 tasks only relate to adding component information to the HTML and JSON formatting. Is it related to https://github.com/MobilityData/gtfs-validator/blob/06578146409e5d844f218be9933dcc84c7cb9064/main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsNetworkSchema.java#L7?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsNetworkSchema.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsNetworkSchema.java:5]: Should be `network.txt`, no?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/runner/ValidationRunner.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/runner/ValidationRunner.java:106]: You can use `optionalValue ` instead of `versionInfo.currentVersion()` on this line

Comment (davidgamez) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/runner/ValidationRunner.java:106]: If this changes the scope of #1640. If not, I suggest adding it as its own PR.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsNetworkSchema.java
Comment (davidgamez) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsNetworkSchema.java:5]: The correct filename is `networks.txt`, `https://gtfs.org/schedule/reference/#networkstxt`

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsNetworkSchema.java
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsNetworkSchema.java:7]: [suggestion]: Add the network schema in a different PR where we can test it in isolation.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareAttributeAgencyIdValidator.java
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/FareAttributeAgencyIdValidator.java:50]: [suggestion]: Have this code in its own PR with similar protecting conditionals added to this PR.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/MatchingFeedAndAgencyLangValidator.java
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/MatchingFeedAndAgencyLangValidator.java:81]: [suggestion]: Have this code in its own PR with similar protecting conditionals added to this PR.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/RouteAgencyIdValidator.java
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/RouteAgencyIdValidator.java:49]: [suggestion]: Have this code in its own PR with similar protecting conditionals added to this PR.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/UrlConsistencyValidator.java
Comment (davidgamez) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/validator/UrlConsistencyValidator.java:122]: [suggestion]: Have this code in its own PR with similar protecting conditionals added to this PR.

**Review by davidgamez — CHANGES_REQUESTED** [review]

**Comment by emmambd** [comment]
@qcdyx Additional changes to scope (could make a new issue/PR for this one): 

Remove the following base components: 

- Route Names
- Agency Information

File: main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsNetworkSchema.java
Comment (qcdyx) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/table/GtfsNetworkSchema.java:5]: Good catch!

**Comment by qcdyx** [comment]
@Emma just fixed the agency problem
image
image

File: main/src/main/java/org/mobilitydata/gtfsvalidator/runner/ValidationRunner.java
Comment (jcpitre) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/runner/ValidationRunner.java:106]: I don't know why it needs its own PR. 
It's simply replacing `versionInfo.currentVersion().get()` with `optionalValue .get()` since we already have optionalValue.
We could also keep it as is it would not be a big deal.

Comment (davidgamez) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/runner/ValidationRunner.java:106]: It has no relation to the scope of the issue. Any bug/side effect on this line will require reverting the commit.

Comment (davidgamez) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/runner/ValidationRunner.java:106]: In this case, it introduces an issue. If the version is not resolved, _gtfsInput_ will be null. The primary process of the validator should not be blocked by the version resolution. The subsequence code consuming the version should support a scenario with no version. I suggest passing the _optionalValue_ as a parameter or using _Optional.orElse(null)_ to pass null avoiding the NoSuchElement exception from _Optional.get_.

Comment (davidgamez) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/runner/ValidationRunner.java:106]: Thanks for the clarification. My point was that the central issue of these few lines is fixing. Currently(master), if the version is not resolved _Optional. get_ throws an exception. The protection _Optional.isPresent_ it's avoiding the exception but skipping validation that results in _a similar_ end-user experience. I suggest ensuring the feed is processed even if the version is not resolved.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/validator/UrlConsistencyValidator.java
Comment (qcdyx) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/validator/UrlConsistencyValidator.java:122]: Actually, the code is not needed once agency.txt is replaced by networks.txt in the GtfsNetworkSchema class. I'll remove them.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/runner/ValidationRunner.java
Comment (qcdyx) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/runner/ValidationRunner.java]: This change is reverted coz it's not needed after the agency.txt is replaced by networks.txt in GtfsNetworkSchema

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (qcdyx) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:230]: Good catch. The List.of method is a list of conditions that each record must satisfy.

**Comment by emmambd** [comment]
@qcdyx Rename _Pathways_ to _Pathways (basic)_

**Comment by qcdyx** [comment]
renamed Pathways to Pathways (basic)
image

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:222]: I don't think these 4 tests (hasMaxSlope, hasMinWidth, hasLength and hasStairCount) go with `Pathway Directions`

Comment (emmambd) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:222]: Correct - should be for 'Pathways (extra)'

**Comment by qcdyx** [comment]
@davidgamez I've made the changes and commented GtfsNetworkSchema code. I'll uncommented the code and remove the requested components in a new PR once this is merged.

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (jcpitre) [inline main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:196]: Is this the right rules for Route-Based Fares? According to the issue, it should be:
`Add Route-Based Fares = If at least 1 value of network_id in routes.txt defined OR 1 record in networks.txt`
I am not sure it was right even _before_ the change.

Comment (emmambd) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java:196]: It does look like we were checking the logic differently from this before. We don't need to check for fare_leg_rule now, just network_id.routes.txt OR networks.txt

**Comment by davidgamez** [comment]
> @davidgamez I've made the changes and commented GtfsNetworkSchema code. I'll uncommented the code and remove the requested components in a new PR once this is merged.

Thanks @qcdyx , Can you kindly remove the commented code from the PR?

File: main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java
Comment (qcdyx) [reply main/src/main/java/org/mobilitydata/gtfsvalidator/report/model/FeedMetadata.java]: Done.

**Review by davidgamez — COMMENTED** [review]
This PR looks good to me. I'll defer to @emmambd and @jcpitre for more deep testing regarding the feature logic.

**Comment by emmambd** [comment]
@qcdyx @davidgamez The example feeds all run correctly - only thing I notice is that Route Names hasn't been deleted. I'll let @jcpitre do the final review + merge :)

**Comment by qcdyx** [comment]
@emmambd I have the route name deletion code locally. I'll commit it in a new PR together with GtfsNetworkSchema.

**Review by jcpitre — APPROVED** [review]
LGTM!

**Pull request merged**
